### ISSUE TITLE
429 Too Many Requests

### ISSUE THREAD

**Issue by chattervast** [craftcms/shopify#28]
Craft CMS: 3.7.28
Shopify: 2.2.0
PHP: 7.4.29

We are experiencing this same issue mentioned in https://github.com/nmaier95/craft-shopify-product-fetcher/issues/26. We have checked the Admin API settings and all of them are set correctly along with the Store Front API settings.
I checked our error logs and we received this error:

`429 Too Many Requests` response:
{"errors":"Exceeded 2 calls per second for api client. Reduce request rates to resume uninterrupted service."}`

Maybe this is part of the issue where the request rate is exceeded and FALSE is returned because no products are returned thus causing the boolean error.
Are the products cached to speed FieldType load?

**Comment by Romanavr** [comment]
We are experiencing the exact same problem.

**Comment by idawgik** [comment]
I just had a client with the same issue. They have a store with 4060 products on it, and trying to load an entry page using the product field would completely fail due to overloaded API.

I did put in a quick and dirty hotfix for our client just adding a `usleep` in the function. I know it's not the best way to address it, but I needed to get things back up for them immediately.

What is ultimately be needed would be a proper rate limiting option in the code. Right now, it's just a function that continually pings the api in a `while` loop with no regard to API rate limits. While this probably works fine on smaller stores, it definitely breaks on stores with larger product catalogs where it has to paginate the requests repeatedly.

**Comment by Romanavr** [comment]
> 

I also recommend optimizing your product request if you haven't already.
For example instead of this
```
        {% for productId in myProducts %}
          {% set shopifyProduct = craft.shopify.getProductById({ id: productId }) %}
          {# ... #}
        {% endfor %}
```
Do something like this
```
        {% set productIds = myProducts|join(',') %}
                   {% set shopifyProducts = craft.shopify.getProducts(
                        {
                            ids: productIds,
                            published_status: 'published',
                        }
                    ).products %}
                   {% for product in shopifyProdicts %}
                   {# ... #}
                   [% endfor %}
```

**Comment by idawgik** [comment]
Thanks, yeah we don't have issues on the frontend. It's purely an issue with loading the field on the entry forms in the admin. I think my wording wasn't the best, so it sounded like I was talking about the frontend entry page.

**Comment by chrishale** [comment]
I've found this to be a problem on a shop with a relatively small amount of products (around 360). Primarily because I'm using this as a custom field inside of a repeatable field. So when a user has added say 5 entries via this repeater/matrix field, then the plugin will end up fetching all products 5 times. It might be useful to have a caching strategy for these scenarios.

**Comment by AugustMiller** [comment]
Hi, all!

Now that the plugin is under Pixel & Tonic’s wing, we've released version 3.0 which includes a number of fixes.

Our new strategy for fetching product data (proactive synchronization into Craft elements) helps avoid connection limits. The Shopify API library should also gracefully handle rate-limiting, so Shopify won't get upset when you synchronize a large storefront! 😜

**Comment by zizther** [comment]
I have encountered this issue when syncing products from Shopify to Craft 4.3.3 using ShopSync 3.0.1
The storefront has a relatively small amount of products.
